go/types/sequence_cursor.go

```diff
@@ -32,16 +62,115 @@ func (cur *sequenceCursor) getItem(idx int) sequenceItem {
 	return cur.seq.getItem(idx)
 }
 
+/*
+ * sync advances the cursor to the next chunk
+ */
 func (cur *sequenceCursor) sync() {
 	d.PanicIfFalse(cur.parent != nil)
 	cur.seq = cur.parent.getChildSequence()
+	cur.readAheadCh = nil
+}
+
+/*
+ * doReadAhead returns true if read-head should be used when reading leaves
+ * of this cursor.
+ *
+ * Only read-ahead if the current sequence is a meta-sequence that contains
+ * leaf sequences.
+ */
+func (cur *sequenceCursor) doReadAhead() bool {
+	if cur.readAhead && cur.readAheadCh == nil {
+		if ms, ok := cur.seq.(metaSequence); ok {
+			return ms.tuples[0].ref.height == 1
+		}
+	}
+	return false
+}
+
+// Records in read-ahead channel
+type readAheadRec struct {
+	idx     int
+	leafSeq sequence
+}
+
+var readAheadParallelism = runtime.NumCPU() * 8
+
+/*
+ * readAheadLeaves initiates read-ahead of child sequences.
+ *
+ * Fills a buffered read-ahead channel with child sequences in ascending index order.
+ * A forward moving cursor reads this channel to find it's next item (see getChildSequence).
+ *
+ * The channel is limited to c * NumCPU. This allows read-ahead to proceed in parallel
+ * while maintaining a bound on memory.
+ */
+func (cur *sequenceCursor) readAheadLeaves() {
+	ms := cur.seq.(metaSequence)
+	startIdx := cur.idx
+	tuplesToRead := len(ms.tuples) - startIdx
+	if tuplesToRead < 2 {
+		// Don't read-ahead if we're starting at the end of a sequence. If the cursor's moving
+		// backward, read-ahead is useless. If the cursor's moving forward, there's no
+		// benefit.
+		cur.readAheadCh = nil
+	} else {
+		parallelism := int(math.Min(float64(readAheadParallelism), float64(tuplesToRead)))
+		input := make(chan interface{}, parallelism)
+
+		cur.readAheadCh = orderedparallel.New(input, func(item interface{}) interface{} {
+			i := item.(int)
+			return &readAheadRec{i, ms.getChildSequence(i)}
+		}, parallelism)
+
+		go func() {
+			for i := startIdx; i < len(ms.tuples); i++ {
+				input <- i
+			}
+			close(input)
+		}()
+	}
 }
 
+/*
+ * getChildSequence retrieves the child at the current cursor position.
+ *
+ * If the the read-ahead channel is enabled, read the child from the channel.
+ */
 func (cur *sequenceCursor) getChildSequence() sequence {
+	if cur.doReadAhead() {
+		cur.readAheadLeaves()
+	}
+	if cur.readAheadCh != nil {
+		FOR: for {
+			select {
+			case item, more := <-cur.readAheadCh:
+				if !more {
```

Comment:
These cases that handle stuff which only happen if the cursor is retreating (a case in which we kinda wouldn't want to be in this loop at all anyway) make me think it might be good to have an explicit readAheadCursor or forwardCursor or something and confine the readahead logic to that.